### DISCUSSION TITLE
Debian Bookworm Kernel with Docker/Podman Support

### DISCUSSION THREAD

**Discussion by inindev** [inindev/rock-5b#10] [General]
> How was it that you are configuring these Kernels now ? Are you using "Stock" Generic Kernel Config ? I had to manually enable LUKS/Dmcrypt/Cryptsetup and, lately, also drivers for Mediatek mt7921au USB-Wireless Adapter.

I have combined the multiple rk3568 and rk3588 repos into a single repo:

**Debian Bookworm 12.5 Image**
https://github.com/inindev/debian-image/releases/tag/v12.5-rc3

The Debian bookworm image for rk3568 devices (nanopi-r5c, nanopi-r5s, odroid-m1, radxa-e25) is just the stock Debian kernel which supports these devices very well.

The Debian bookworm image for rk3588 devices (nanopc-t6, orangepi-5, orangepi-5-plus, rock-5b) uses the Debian sid kernel (currently 6.8.9) which supports the rk3588 devices very well.

The rk3588 devices have an ```apt``` directive in the file ```/etc/apt/preferences.d/kernel``` file to update all packages based on bookworm and only the kernel from sid:

```
Package: *
Pin: release n=bookworm
Pin-Priority: 900

Package: linux-image-arm64
Pin: release n=sid
Pin-Priority: 1001
```

This will keep the image updating from Debian upstream.  The nice thing about this approach is that everything will remain up-to-date using ```apt update``` and ```apt upgrade``` commands and support the wide breadth of modules Debian compiles into their kernels.

**Comment by inindev** [comment]
PS: u-boot and a custom kernel are available for those who are interested:

**u-boot v2024.07-rc3**
https://github.com/inindev/uboot-rockchip/releases/tag/v2024.07-rc3

**linux kernel v6.9.2-1 arm64 inindev**
https://github.com/inindev/linux-rockchip/releases/tag/v6.9.2

The custom 6.9.2 kernel picks up usb and hdmi patches for the rock-5b and nanopc-t6 from collabora: https://gitlab.collabora.com/hardware-enablement/rockchip-3588/linux

**Comment by luckylinux** [comment]
@inindev: Sorry for my late Reply to your Comments (also in the original Issue).

The Kernel Update is probably a very good thing to have, as I'm currently running into some issues with the Mediatek mt7921au Driver crashing periodically/randomly.

However, due to the other Issue with Debian, I think (in the long Term) I'll have to flash Fedora AARCH64 on these boards, since Podman 5.x / Pasta-Passt doesn't really work well on Debian Bookworm (neiter on Trixie/Testing nor on Sid/Unstable for that matter) :cry:.

Too bad because I really like Debian but Podman Compatibility is just ... well ... not that great, so I'll be hopefully migrating everything to Fedora Server/Minimal.

**Comment by inindev** [comment]
One other thing to consider is that linux kernel 6.10 is about to move to stable which should push sid from 6.9 to 6.10 in a week or two:
https://packages.debian.org/sid/kernel/linux-image-arm64

**Reply by luckylinux** [reply]
Linux 6.10 is now Mainline (just checked since I'm still currently running 6.6.x and will upgrade to 6.6.41 Today).

Nothing to do with you (I really liked your Project !) or Debian (which I basically run everywhere else), unfortunately my move to Fedora is purely dictated by the requirement of Podman 5.x and Pasta-Passt Networking Stack :cry:. SELinux is definitively NOT fun (for one Thing) ...

But yeah, it will still take a while before I get to that "Conversion" ... 1000 other Things to be done :laughing:.

**Comment by MartinHerren** [comment]
For info, i still run the Debian Trixie image generated from this repo, in the mean time i switched to the native aarch64 kernel from Debian.

Currently it is 6.9.12, previously 6.9.10, without an issue, including docker support.

The one thing that seems missing is the support for the QSPI (probably missing in the device tree), so not possible to update u-boot. There might be other things not supported, haven't identified any yet.
Otherwise my Rock 5B runs 24/24 and 7/7 on the stock Debian kernel now without an issue. Installation is on NVMe with u-boot on QSPI.

**Reply by luckylinux** [reply]
Seems like I'm also hit by the SPI not getting recognized, although I am on Kernel 6.6.41.

I'm trying to switch to Fedora (to get Podman 5.x) but I cannot flash EDK2 Bootloader into the SPI since `/dev/mtdblk0` or `/dev/mtd*` for that Matter don't show up :disappointed:.

So basically I'm stuck ...

EDIT 1: I see Podman 5.0.3 is in Debian Testing/Trixie now, but Fedora already has Podman 5.2.0. I cannot do the Testing -> Stable APT Polling though anymore since Libc6 / time64t migration already occurred in Debian Trixie. Anyways most likely `pasta`/`passt` still have a quite old Version compared to what I need ...

EDIT 2: Maybe I need to flash the Official Images to an SDCard and boot from it to be able to flash SPI ?

**Reply by MartinHerren** [reply]
If you still have a bootable kernel from this repo (or able to reinstall one from the .deb) you can reboot on it to get the /dev/mtdblk*s back.

**Reply by inindev** [reply]
I have not used the EDK2 Bootloader but if the Linux Kernel is loaded, you are well past the bootloader, no?

In any event, here are two kernels worth looking at:

**collabora**
https://gitlab.collabora.com/hardware-enablement/rockchip-3588/linux/-/artifacts
https://gitlab.collabora.com/hardware-enablement/rockchip-3588/linux/-/jobs/364572/artifacts/download?file_type=archive

**debian**
https://packages.debian.org/sid/kernel/linux-image-arm64

**Reply by luckylinux** [reply]
@inindev Yes, but the Issue with the patched Kernel (at least the 6.6.x Series that I'm still using) is that `/dev/mtdblk0` or `/dev/mtd*` don't show up so I cannot flash the EDK2 Bootloader in order to install Fedora. Again, it's not that I don't like Debian, just that I need Podman 5.2.x and (updated) Podman Support in Debian is like ... meh :disappointed:.

I could probably try out the Official Kernel 5.x (?) with an Official Image provided by Radxa for the Rock 5B just to do the flashing though.

I can see however that `podman-5.2.1+ds1-2` just landed in Debian Testing / Trixie, together with `podman-compose-1.2.0-2` and `passt-0.0~git20240821.1d6142f-1` . Maybe that makes the decision more complicated now since that (in Theory) would be what I need :rofl:.

But given my previous Experiences with running Debian Testing (or Unstable for that Matter) and dependency resolution during updates, it's not really a smooth Experience. Not on Desktops at least (maybe on Servers it's better but who runs Debian Testing on a Server ?).

**Reply by luckylinux** [reply]
@MartinHerren I am curious since now I also got Kernel 6.10.6 from Backports installed. USB 2.0 does NOT work at all and USB 3.0 keeps logging to `dmesg` this: `[  563.269511] usb usb2-port1: Cannot enable. Maybe the USB cable is bad?` (whether I use a Cable Extension or not).

The FENVI USB WiFi Adapter (based on Mediatek MT71xx Chipset) is NOT getting recognized anymore.

**Comment by inindev** [comment]
The debian image is setup to be 100% Debian software now.
debian bookworm v12.6-2 arm64
https://github.com/inindev/debian-image/releases

This image uses the sid kernel for rk3588 devices.
https://packages.debian.org/sid/kernel/linux-image-arm64

The sid kernel is selected here:
/etc/apt/preferences.d/99-trixie-kernel

Support for video and usb is still on its way upstream and still requires a custom kernel.  I just built the 6.11 kernel yesteray:
https://github.com/inindev/linux-rockchip/releases

To use the inindev 6.11 kernel:
```
wget https://github.com/inindev/linux-rockchip/releases/download/v6.11.0/linux-image-6.11.0-1-arm64_6.11.0-1_arm64.deb
sudo dpkg -i linux-image-6.11.0-1-arm64_6.11.0-1_arm64.deb
sudo reboot
```
The inindev 6.11 kernel is patched to have working usb and video.  It should also support podman unless I broke that when I tried cleaning up unneeded kernel packages about 6 mo. ago.
